### PR TITLE
probeservices, oonimkall: implement checkin api

### PR DESCRIPTION
**Pull request by lorenzoPrimi**
Fixes https://github.com/ooni/probe-engine/issues/1148

### CODE REVIEW

File: probeservices/checkin.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+	V     int               `json:"v"`
+}
+
+// CheckIn pobes ask for tests to be run or otherwise go back to sleep
+// The config argument contains the mandatory settings.
+// Returns the list of tests to run and the URLs, on success, or an explanatory error, in case of failure.
+func (c Client) CheckIn(ctx context.Context, config model.CheckInConfig) (model.CheckInInfo, error) {
+	var response checkInResult
+	err := c.Client.PostJSON(ctx, "/api/v1/check-in", config, &response)
+	if err != nil {
```
Comment (bassosimone) [inline probeservices/checkin.go:20]:
```suggestion
	if err := c.Client.PostJSON(ctx, "/api/v1/check-in", config, &response); err != nil {
```

File: model/checkininfo.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package model
+
+// CheckInInfoWebConnectivity contains the array of URLs returned by the checkin API
+type CheckInInfoWebConnectivity struct {
+	URLs []URLInfo `json:"urls"`
```
Comment (bassosimone) [inline model/checkininfo.go:5]: IIRC there is also the reportID we should be using. I would like to add this to the structure even though we are currently not using it?

File: probeservices/checkin_test.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	client := newclient()
+	client.BaseURL = "https://ams-pg-test.ooni.org"
+	config := model.CheckInConfig{
+		Charging:        true,
+		OnWiFi:          true,
+		Platform:        "string",
+		ProbeASN:        "string",
+		ProbeCC:         "string",
+		RunType:         "string",
+		SoftwareVersion: "string",
```
Comment (bassosimone) [inline probeservices/checkin_test.go:21]: Can we use realistic values here?

File: probeservices/checkin_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		WebConnectivity: model.CategoryCodes{
+			CategoryCodes: []string{"NEWS", "CULTR"},
+		},
+	}
+	ctx := context.Background()
+	result, err := client.CheckIn(ctx, config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	//TODO what do we want to check?
```
Comment (bassosimone) [inline probeservices/checkin_test.go:31]:
```suggestion
	// TODO(lorenzoPrimi): what do we want to check?
```

Comment (bassosimone) [reply probeservices/checkin_test.go:31]: Apart from this style-oriented comment, I think checking that the category codes are okay is already good. I also think we should be checking whether we have got a report ID.

File: model/checkininfo.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+}
+
+// CheckInInfo contains the return test objects from the checkin API
+type CheckInInfo struct {
+	WebConnectivity CheckInInfoWebConnectivity `json:"web_connectivity"`
```
Comment (bassosimone) [inline model/checkininfo.go:10]: My understanding is that this property is optional. If that is the case, then we should use instead:

```suggestion
	WebConnectivity *CheckInInfoWebConnectivity `json:"web_connectivity"`
```

This also means that we need to be careful when manipulating the returned value. This is a sad side effect of languages that allow for `nil` values and we need to live with that.

File: probeservices/checkin.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+type checkInResult struct {
+	Tests model.CheckInInfo `json:"tests"`
+	V     int               `json:"v"`
+}
+
+// CheckIn pobes ask for tests to be run or otherwise go back to sleep
+// The config argument contains the mandatory settings.
+// Returns the list of tests to run and the URLs, on success, or an explanatory error, in case of failure.
+func (c Client) CheckIn(ctx context.Context, config model.CheckInConfig) (model.CheckInInfo, error) {
```
Comment (bassosimone) [inline probeservices/checkin.go:17]:
```suggestion
func (c Client) CheckIn(ctx context.Context, config model.CheckInConfig) (*model.CheckInInfo, error) {
```

I think we want to return either a valid pointer or an error. This is a minor thing, but I actually happen to quite like this either-or semantics where we basically emulate the returning of a sum type.

After this change, tests need to be updated.

File: model/checkinconfig.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package model
+
+// CategoryCodes is an array of category codes
+type CategoryCodes struct {
+	CategoryCodes []string `json:"category_codes"` // CategoryCodes is an array of category codes
+}
```
Comment (bassosimone) [inline model/checkinconfig.go:6]: Maybe I'm wrong, but I thought we already had a structure for doing this? If we do not, then I suggest it may be a good idea to call this structure `CheckInCategoryCodes`.

File: probeservices/checkin.go (hunk 1 above)
Comment (bassosimone) [reply probeservices/checkin.go:20]: I like this style more, because it's more compact.

File: probeservices/checkin_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	if len(result.WebConnectivity.URLs) < 1 {
```
Comment (bassosimone) [inline probeservices/checkin_test.go:32]: This needs to check for `result.WebConnectivity == nil` once we make this field nullable.

File: probeservices/checkin_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+			CategoryCodes: []string{"NEWS", "CULTR"},
+		},
+	}
+	ctx := context.Background()
+	result, err := client.CheckIn(ctx, config)
+	if err == nil || !strings.HasSuffix(err.Error(), "invalid control character in URL") {
+		t.Fatal("not the error we expected")
+	}
+	//TODO how to do a proper check here?
+	if len(result.WebConnectivity.URLs) != 0 {
```
Comment (bassosimone) [inline probeservices/checkin_test.go:63]: On failure, we want `result` to be `nil`. I think there is one extra case to test where the server returns us a valid but empty WebConnectivity. We can either provoke this case using the real server or using mocking.

**Review by bassosimone — COMMENTED** [review]
Thank you for working on this! This code is already a very robust base to apply small incremental changes on! I have provided suggestions on how to adapt the API a little bit and make the tests more robust.

File: probeservices/checkin_test.go (hunk 4 above)
Comment (lorenzoPrimi) [reply probeservices/checkin_test.go:31]: I don't understand this report_id for what is needed, and we are not gonna use it anytime soon. Do we need the test now?

File: probeservices/checkin_test.go (hunk 3 above)
Comment (lorenzoPrimi) [reply probeservices/checkin_test.go:21]: Platform:        "android",
		ProbeASN:        "AS12353",
		ProbeCC:         "PT",
		RunType:         "?",
		SoftwareVersion: "2.7.1",

Comment (lorenzoPrimi) [reply probeservices/checkin_test.go:21]: I don't know what runtype is. @FedericoCeratto ?

File: model/checkinconfig.go (hunk 7 above)
Comment (lorenzoPrimi) [reply model/checkinconfig.go:6]: There is no other CategoryCodes struct

File: probeservices/checkin_test.go (hunk 3 above)
Comment (lorenzoPrimi) [reply probeservices/checkin_test.go]: solved

File: model/checkinconfig.go (hunk 7 above)
Comment (bassosimone) [reply model/checkinconfig.go:6]: So, I suggest either you put it into `model/categorycodes.go` or prefix `CheckIn` to it?

Comment (bassosimone) [reply model/checkinconfig.go:6]: Now that I take another look at the code, I believe we should just have `model/checkin.go` with everything inside it using the `CheckIn` prefix to make it super easy to understand where the code lives!

File: model/checkininfo.go (hunk 2 above)
Comment (bassosimone) [reply model/checkininfo.go:5]: Let us put all in `model/checkin.go` so this can become `CheckInWebConnectivity`!

File: probeservices/checkin.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+	"github.com/ooni/probe-engine/model"
+)
+
+type checkInResult struct {
+	Tests model.CheckInInfo `json:"tests"`
+	V     int               `json:"v"`
+}
+
+// CheckIn pobes ask for tests to be run or otherwise go back to sleep
```
Comment (bassosimone) [inline probeservices/checkin.go:14]: I think we need to spell check this sentence!

**Review by bassosimone — COMMENTED** [review]
Did another pass!

**Comment by bassosimone** [comment]
Remaining TODOs:

- [ ] make sure we generate the desired interfaces for Android/iOS
- [ ] re-read the whole diff
- [ ] make sure we're testing everything we should be testing

**Review by bassosimone — APPROVED** [review]
:whale:
